codegen/ace/src/circuit.rs

```diff
@@ -0,0 +1,507 @@
+
+        while let Some(id) = self.process.pop() {
+            let op = self.circuit[id];
+            if self.processed.contains(&id) {
+                return Some((id, op));
+            } else {
+                self.process.push(id);
+                self.visited.insert(id);
+                self.processed.insert(id);
+                match op {
+                    Value(_) => {}
+                    Add(l, r) | Sub(l, r) | Mul(l, r) => {
+                        if !self.visited.contains(&r) {
+                            self.visited.insert(r);
+                            self.process.push(r)
+                        };
+                        if !self.visited.contains(&l) {
+                            self.visited.insert(l);
+                            self.process.push(l)
+                        };
+                    }
+                };
+            }
+        }
+        None
+    }
+}
+
+/// Remap the children of an operation. Typically when migrating an operation
+/// between data structures.
+pub fn remap_op(op: Operation, mapping: &BTreeMap<NodeIndex, NodeIndex>) -> Operation {
+    use Operation::*;
+    match op {
+        Value(_) => op,
+        Add(l, r) => {
+            let l = mapping[&l];
+            let r = mapping[&r];
+            Add(l, r)
+        }
+        Sub(l, r) => {
+            let l = mapping[&l];
+            let r = mapping[&r];
+            Sub(l, r)
+        }
+        Mul(l, r) => {
+            let l = mapping[&l];
+            let r = mapping[&r];
+            Mul(l, r)
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct CircuitBuilder {
+    // The ACE circuit being built
+    circuit: Circuit,
+    // A cache of nodes already inserted in the circuit, used to avoid duplicates
+    cache: BTreeMap<Operation, NodeIndex>,
+    // Number of inputs used so far
+    n_inputs: usize,
+    // Constants zero and one are inserted first so that they can be used as
+    // identity and absorbing element in addition and multiplication.
+    pub zero: NodeIndex,
+    pub one: NodeIndex,
+    // Stores the position of the alpha input once it's set.
+    pub alpha: Option<NodeIndex>,
+    // Tracks the next power of alpha to be used. Starts at 1=alpha^0.
+    next_alpha: NodeIndex,
+}
+
+impl CircuitBuilder {
+    pub fn new() -> Self {
+        let mut cache: BTreeMap<Operation, NodeIndex> = BTreeMap::new();
+        let mut circuit = Circuit::new();
+        let constants: Vec<NodeIndex> = [0, 1]
+            .iter()
+            .map(|c| {
+                let op = Operation::Value(Value::Constant(*c as u64));
+                let out = circuit.append(op);
+                cache.insert(op, out);
+                out
+            })
+            .collect();
+        Self {
+            cache,
+            circuit,
+            n_inputs: 0,
+            zero: constants[0],
+            one: constants[1],
+            alpha: None,
+            next_alpha: constants[1],
+        }
```

Comment:
Arrays have a `map` method that returns an array of the same size. It can be destructured in-place. This avoids an extra `Vec` allocation
```suggestion
        let [zero, one] = [0, 1].map(|c| {
            let op = Operation::Value(Value::Constant(c as u64));
            let out = circuit.append(op);
            cache.insert(op, out);
            out
        });
        Self {
            cache,
            circuit,
            n_inputs: 0,
            zero,
            one,
            alpha: None,
            next_alpha: one,
        }
```

Reply:
Very nice. Done.